finality-aleph/src/data_io.rs

```diff
@@ -1,13 +1,190 @@
-use crate::{Error, Metrics};
+use crate::{network::AlephNetworkData, Error, Metrics};
 use aleph_bft::OrderedBatch;
 use futures::channel::{mpsc, oneshot};
-use log::debug;
+use lru::LruCache;
 use parking_lot::Mutex;
 use sp_consensus::SelectChain;
 use sp_runtime::traits::{Block, Header};
-use std::{sync::Arc, time::Duration};
+use std::{
+    collections::{hash_map::Entry, HashMap, HashSet},
+    marker::PhantomData,
+    sync::Arc,
+    time::Duration,
+};
 
 const REFRESH_INTERVAL: u64 = 100;
+use futures::channel::mpsc::{UnboundedReceiver, UnboundedSender};
+use futures_timer::Delay;
+use log::{debug, trace};
+use sc_client_api::{backend::Backend, BlockchainEvents};
+use sp_runtime::generic::BlockId;
+use tokio::stream::StreamExt;
+
+type MessageId = u64;
+const AVAILABLE_BLOCKS_CACHE_SIZE: usize = 1000;
+const PENDING_MESSAGE_LIFETIME: MessageId = 100_000;
```

Comment:
the naming here is a bit confusing, is it an id or a duration?

Reply:
Yes, it isn't really a lifetime. I changed it to `MESSAGE_ID_BOUNDARY` as this is a boundary for oldest message id that should be kept in pending messages.